crates/rari-doc/src/redirects.rs

```diff
@@ -99,3 +113,721 @@ pub fn resolve_redirect(url: &str) -> Option<Cow<'_, str>> {
+                parsed_url.scheme()
+            )));
+        }
+    } else if url.starts_with('/') {
+        // Internal URL, perform validations
+
+        check_url_invalid_symbols(url)?;
+        validate_url_locale(url)?;
+
+        // Split by '#', take the bare URL
+        let bare_url = url.split('#').next().unwrap_or("");
+
+        let parts: Vec<&str> = bare_url.split('/').collect();
+        if parts.len() < 2 {
+            return Err(DocError::InvalidRedirectToURL(format!(
+                "To-URL '{}' does not have enough parts for locale validation.",
+                url
+            )));
+        }
+
+        let to_locale = parts[1];
+        if to_locale != locale.as_url_str().to_lowercase() {
+            // Different locale, no need to check path
+            return Ok(());
+        }
+
+        let (path, _, _, _) = url_path_to_path_buf(&url)?;
+        let path = root_for_locale(*locale)?
+            .join(locale.as_folder_str())
+            .join(path);
+        if !path.exists() {
+            return Err(DocError::InvalidRedirectToURL(format!(
+                "To-URL '{}' resolves to a non-existing file/folder at '{}'.",
+                url,
+                path.display()
+            )));
+        }
+    } else {
+        return Err(DocError::InvalidRedirectToURL(format!(
+            "To-URL '{}' has to be external (https://) or start with '/'.",
+            url
+        )));
+    }
+
+    Ok(())
+}
+
+fn validate_url_locale(url: &str) -> Result<(), DocError> {
+    let parts: Vec<&str> = url.split('/').collect();
+    if parts.len() < 3 {
+        return Err(DocError::InvalidRedirectToURL(format!(
+            "To-URL '{}' does not have enough parts for locale validation.",
+            url
+        )));
+    }
+
+    let to_locale = parts[1];
+    if !Locale::for_generic_and_spas()
+        .iter()
+        .map(|l| l.as_url_str().to_lowercase())
+        .collect::<HashSet<String>>()
+        .contains(&to_locale.to_lowercase())
+    {
+        return Err(DocError::InvalidRedirectToURL(format!(
+            "Locale prefix '{}' in To-URL '{}' is not valid.",
+            to_locale, url
+        )));
+    }
+
+    Ok(())
+}
+
+fn is_vanity_redirect_url(url: &str) -> bool {
+    let locale_urls = Locale::for_generic_and_spas()
+        .iter()
+        .map(|l| format!("/{}/", l.as_url_str()))
+        .collect::<HashSet<String>>();
+    return locale_urls.contains(&url.to_string());
+}
+
+fn check_url_invalid_symbols(url: &str) -> Result<(), DocError> {
+    for symbol in FORBIDDEN_URL_SYMBOLS.iter() {
+        if url.contains(symbol) {
+            let escaped_symbol = match symbol.as_str() {
+                "\n" => "\\n",
+                "\t" => "\\t",
+                _ => symbol,
+            };
+            return Err(DocError::InvalidRedirectToURL(format!(
+                "URL '{}' contains forbidden symbol '{}'.",
+                url, escaped_symbol
+            )));
+        }
+    }
+    Ok(())
```

Comment:
Sorry for the drive by comment. I was looking at merge conflicts. This would save the LazyLock (Well that could be replaces by `['\t', '\n']` anyway). 
```suggestion
    if let Some(symbol_index) = url.find(['\t', '\n']) {
        if let Some(escaped_symbol) = &url[symbol_index..]
            .chars()
            .next()
            .map(|c| c.escape_default())
        {
            return Err(DocError::InvalidRedirectToURL(format!(
                "URL '{url}' contains forbidden symbol '{escaped_symbol}'."
            )));
        }
    }
```